ddtrace/internal/processor.py

```diff
@@ -12,11 +13,71 @@
 
 
 @attr.s
-class TraceProcessor(object):
+class Stitcher(object):
+    _partial_flush_enabled = attr.ib(type=bool)  # type: bool
+    _partial_flush_min_spans = attr.ib(type=int)  # type: int
+    _traces = attr.ib(factory=dict, init=False)  # type: List[_Trace]
+
+    @attr.s
+    class _Trace(object):
+        spans = attr.ib(default=attr.Factory(list))  # type: List[Span]
+        num_finished = attr.ib(type=int, default=0)  # type: int
+
+    def _get_or_create_trace(self, trace_id):
+        # type: (int) -> Stitcher._Trace
+        if trace_id in self._traces:
+            return self._traces[trace_id]
+        else:
+            trace = self._Trace()
+            self._traces[trace_id] = trace
+            return trace
+
+    def add_span(self, span):
+        # type: (Span) -> None
+        trace = self._get_or_create_trace(span.trace_id)
+        trace.spans.append(span)
+
+    def finish_span(self, span):
+        # type: (Span) -> Optional[List[Span]]
+        trace = self._get_or_create_trace(span.trace_id)
+        trace.num_finished += 1
+        if trace.num_finished == len(trace.spans) or (
+            self._partial_flush_enabled and trace.num_finished >= self._partial_flush_min_spans
+        ):
+            finished = [s for s in trace.spans if s.finished]
+            trace.spans = [s for s in trace.spans if not s.finished]
+            trace.num_finished -= len(finished)
+
+            if len(trace.spans) == 0:
+                del self._traces[span.trace_id]
+            return finished
+        return None
+
+
+@attr.s(eq=False)
+class SpanProcessor(object):
```

Comment:
We went from something pretty generic so something ultra-specific. I feel like we lost all the abstraction we made from the `Processor` approach to split back the current code in different classes.

`process` was great as it was stateless.

Now there are 5 different mechanisms at play at different stages:
1. `on_start_span`
2. `on_span_finish`
3.  `stitcher` in `on_span_finish`
4. `filter` inside `on_span_finish`
5. `sample` inside `on_span_finish`

I know I'm a bit delusional but I'd hope we could design in a more modular approach rather than getting back at "everything like before but now it's in more classes" 😁

Reply:
I think 1. and 2. make sense to define the SpanProcessor API.

Unfortunately, it does suffer from the existing constraints: piecing the traces together is going to require state; filters are applied to entire traces (not just individual spans); sampling decisions are made based off whether any span in the trace has been sampled (again needs the whole trace). I tried not to change the unrelated code around too much to avoid any regressions but I think that we can and should do better than just copying it from one place to another.

Ideally the filtering and sampling could happen in different processors so I will give that a shot!

Reply:
Gave it a shot, let me know what you think @jd. I've split out a big part of it in #2210 to lay the groundwork if we're happy with this approach.